compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -1583,15 +1592,33 @@ impl<'hir> LoweringContext<'_, 'hir> {
         let (iter_pat, iter_pat_nid) =
             self.pat_ident_binding_mode(head_span, iter, hir::BindingAnnotation::MUT);
 
-        // `match Iterator::next(&mut iter) { ... }`
         let match_expr = {
             let iter = self.expr_ident(head_span, iter, iter_pat_nid);
-            let ref_mut_iter = self.expr_mut_addr_of(head_span, iter);
-            let next_expr = self.expr_call_lang_item_fn(
-                head_span,
-                hir::LangItem::IteratorNext,
-                arena_vec![self; ref_mut_iter],
-            );
+            let next_expr = if is_await {
+                // `async_iter_next(unsafe { Pin::new_unchecked(&mut iter) }).await`
```

Comment:
Please split this up across the several `let` exprs below for clarity of what each line is doing.